Deploy Automation: Add prepare-release command

markbiek opened the pull request:
### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Related to #1096
Closes #1099

This command rolls the following steps into a single command:

1. Bump version
2. Update changelog
3. Optionally download translation files
4. Create a PR

<!-- The next section is mandatory. If your PR doesn't require testing, please indicate that you are purposefully omitting instructions. -->

- [ ] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Otherwise, please include detailed instructions on how these changes can be tested. Please, make sure to review and follow the guide for writing high-quality testing instructions below. -->

- [x] Have you followed the [Writing high-quality testing instructions guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions)?

#### ** Preparing for testing **
1. Ensure you're running Node v16.
2. Run `npm install` in the project root.

#### Normal execution

1. Run `npm run prepare-release`.
3. You should be guided through the process of incrementing the version and updating the readme.
4. When prompted to download translations, you can say yes or now. You may see an error if you say yes but the script will continue running normally.
5. Say yes to the prompts to create a new PR and to open the PR in your browser.
6. You should see the following messages and the PR should open in your browser.

![image](https://github.com/Automattic/wc-calypso-bridge/assets/917632/9f7c2342-2720-4610-8fa8-91a758ede5aa)

#### Release branch already exists locally
This assumes you have already run "Normal execution" but _haven't_ done "Test cleanup" yet.

1. Run `npm run prepare-release`.
2. Follow the prompts to increment the version.
3. You should get the following error and be switched back to the original branch you were on.
![image](https://github.com/Automattic/wc-calypso-bridge/assets/917632/2620e33e-407c-4bb2-887f-b59e34bd5d25)

#### Release branch exists on Github
This assumes you have already run "Normal execution" and "Release branch already exists locally" but _haven't_ done "Test cleanup" yet.

1. Run `git branch -D update-version-X.X.X ` to delete the branch locally. The branch should still exist on Github.
2. Run `npm run prepare-release`.
3. Follow the prompts until you're prompted to create the pull request. Say yes to create the pull request. You should get an error like the below.
![image](https://github.com/Automattic/wc-calypso-bridge/assets/917632/1c3f6632-0e0f-4ffa-90ef-28a4f19559a2)
4. If you say yes, it should continue creating the pull request normally.

#### **Test Cleanup***
When you're done testing:

1. Add a comment like "Fake PR for testing." to the PR created above and then close the PR.
5. Delete the release branch by running `git branch -D update-version-X.X.X && git push --delete origin update-version-X.X.X`.

<!-- End testing instructions -->

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

markbiek committed:
Add prepare-release command

This command rolls the following steps into a single command:

1. Bump version
2. Update changelog
3. Optionally download translation files
4. Create a PR

markbiek commented:
@andres-blanco This should be ready to test again when you have some time.

andres-blanco commented:
@markbiek I was trying again and got this message:
<img width="996" alt="image" src="https://github.com/Automattic/wc-calypso-bridge/assets/375980/e3ed0b19-2361-46ac-b317-b96dd85ba104">
It's a little confusing. If I need to run `git push --set-upstream origin update-version-2.1.7` why is it warning me that it will try force pushing?

markbiek commented:
@andres-blanco I'm having trouble reproducing that issue you posted above. Is that still happening for you and can you give me any more details?

markbiek commented on scripts/prep-release.js:
```diff
@@ -0,0 +1,150 @@
+	// Push our release prep branch
+	const git = gitFactory();
+	try {
+		await git.push( 'origin', git.branch() );
```
That makes sense. I must not be having that issue because I had that same branch at one point before.

Thanks!

valterlorran replied:
To reproduce it, you must delete your test branches locally and remotely.

valterlorran replied:
Sorry! I missed your comment above, that's it :)

markbiek committed:
Update scripts/prep-release.js

Co-authored-by: valterlorran <valterlorran@hotmail.com>

markbiek commented:
@andres-blanco @valterlorran I've added Valters fix if either of you want to try testing again.

valterlorran commented on scripts/prep-release.js:
```diff
@@ -0,0 +1,150 @@
+			process.exit( 1 );
+		}
+
+		await git.push( 'origin', git.branch(), { '--force': null } );
```
I think we should also handle errors in the case to revert it to the original branch.

valterlorran commented on scripts/prep-release.js:
```diff
@@ -0,0 +1,150 @@
+	// Push our release prep branch
+	const git = gitFactory();
+	try {
+		await git.push( 'origin', git.branch(), [ '--set-upstream' ] );
```
I think this maybe can solve the problem I am facing.
```suggestion
		await git.branch(['--set-upstream-to=origin/${branchName}', branchName]);
		await git.push( 'origin', git.branch() );
```

valterlorran commented on scripts/prep-release.js:
```diff
@@ -0,0 +1,150 @@
+		await git.push( 'origin', git.branch(), [ '--set-upstream' ] );
+	} catch ( e ) {
+		error(
+			`Something went wrong pushing ${ git.branch() } to origin.\n${ e }`
```
I think the `git.branch()` function is returning an object:
<img width="465" alt="Screen Shot 2023-06-07 at 10 17 55" src="https://github.com/Automattic/wc-calypso-bridge/assets/1234758/da325f91-37d7-4589-8b25-3e99ef48fd96">

andres-blanco commented:
> @andres-blanco @valterlorran I've added Valter's fix if either of you want to try testing again.

Mark, I've just retried and I get the same message
<img width="998" alt="image" src="https://github.com/Automattic/wc-calypso-bridge/assets/375980/7e2ac9a3-4e48-420e-bbdf-59be8c048f8a">

markbiek committed:
Update scripts/prep-release.js

Co-authored-by: valterlorran <valterlorran@hotmail.com>

markbiek committed:
Fix issues pushing the release branch

markbiek commented:
@valterlorran @andres-blanco I think this is working now. I've tried a couple of different times, updating both the patch and minor versions and I was able to generate the PR.

Let me know if you're still having trouble.

ilyasfoo commented:
Thanks for your work here, @markbiek! I tried this with releasing `2.1.7` and it seems to work pretty smoothly. The only issue I had was `wp i18n make-pot` ended up with an error related to a string not having `translators:` comment to assist translators. I don't think it's not something we have to address in this PR.

Also, regarding changelogs, I think it may be a bit tricky for releaser to be managing it by picking PRs. I think if we can adopt [Jetpack changelogger](https://github.com/Automattic/jetpack-changelogger) or its workflow, it'll be much more streamlined. Individual PR authors would be responsible for creating changelog and releaser would just need to compile them.

markbiek commented:
@ilyasfoo I think using the jetpack changelogger is a great idea but I worry that it's too big of a change to include with this diff.

What do you think about me switching the readme update here to just a simple text entry for now so we can get this PR out the door? Then we can implement the jetpack thing in a separate PR.

ilyasfoo commented:
Sounds good, @markbiek!

markbiek committed:
Rename exception handling vars

The exception variable was colliding with the error function

markbiek committed:
Switch to free-text entry for changelog updates

Manually selecting commits has been a little tricky/unreliable. This
switches to launching the editor so the user can update the changelog
manually.

markbiek commented:
@ilyasfoo I've updated this so now it launches the editor and lets the user enter their changelog updates manually.

ilyasfoo reviewed (approved):
Thanks, @markbiek! LGTM!
